### PR TITLE
Don't add -Wl,--no-undefined for clang, windows 10 [SDL2]

### PR DESCRIPTION
**Pull request by AlpRabbit**
Since clang on Windows 10 doesn't have `-Wl,--no-undefined` option, disabling it. [Same thing is done in SDL](https://github.com/libsdl-org/SDL/commit/34ecd71e80b651c9840f60d91be60e30305fdc6f).

### CODE REVIEW

**Comment by AlpRabbit** [comment]
This is `CMakeConfigureLog.yaml`
```yaml
kind: "try_compile-v1"
    backtrace:
      - "C:/Program Files/CMake/share/cmake-3.26/Modules/Internal/CheckSourceCompiles.cmake:101 (try_compile)"
      - "C:/Program Files/CMake/share/cmake-3.26/Modules/CheckCSourceCompiles.cmake:76 (cmake_check_source_compiles)"
      - "subproject/SDL_image/cmake/PrivateSdlFunctions.cmake:257 (check_c_source_compiles)"
      - "subproject/SDL_image/cmake/PrivateSdlFunctions.cmake:266 (sdl_check_linker_flag)"
      - "subproject/SDL_image/CMakeLists.txt:291 (sdl_target_link_options_no_undefined)"
    checks:
      - "Performing Test HAVE_WL_NO_UNDEFINED"
    directories:
      source: "C:/Users/RYOSU/Projects/sdltest/build/CMakeFiles/CMakeScratch/TryCompile-iy6n6a"
      binary: "C:/Users/RYOSU/Projects/sdltest/build/CMakeFiles/CMakeScratch/TryCompile-iy6n6a"
    cmakeVariables:
      CMAKE_C_FLAGS: ""
      CMAKE_MODULE_PATH: "C:/Users/RYOSU/Projects/sdltest/subproject/SDL_image/cmake"
      CMAKE_POSITION_INDEPENDENT_CODE: "ON"
    buildResult:
      variable: "HAVE_WL_NO_UNDEFINED"
      cached: true
      stdout: |
        Change Dir: C:/Users/RYOSU/Projects/sdltest/build/CMakeFiles/CMakeScratch/TryCompile-iy6n6a
        
        Run Build Command(s):C:/PROGRA~1/ninja/bin/ninja.exe -v cmTC_99fd9 && [1/2] C:\\PROGRA~1\\mingw64\\bin\\gcc.exe -DHAVE_WL_NO_UNDEFINED   -o CMakeFiles/cmTC_99fd9.dir/src.c.obj -c C:/Users/RYOSU/Projects/sdltest/build/CMakeFiles/CMakeScratch/TryCompile-iy6n6a/src.c
        [2/2] cmd.exe /C "cd . && C:\\PROGRA~1\\mingw64\\bin\\gcc.exe  -Wl,--no-undefined CMakeFiles/cmTC_99fd9.dir/src.c.obj -o cmTC_99fd9.exe -Wl,--out-implib,libcmTC_99fd9.dll.a -Wl,--major-image-version,0,--minor-image-version,0  -lkernel32 -luser32 -lgdi32 -lwinspool -lshell32 -lole32 -loleaut32 -luuid -lcomdlg32 -ladvapi32 && cd ."
        
      exitCode: 0
```

**Comment by madebr** [comment]
Thanks for the log, but the log is for a build with the `gcc.exe` compiler, not clang.
gcc was able to link the test program without anything printed to stdout/stderr.

**Comment by madebr** [comment]
Can you show us the warning/error you want to fix here? Including the command that caused it. Do the build in verbose mode for that: `cmake --build . --verbose`.

**Comment by sezero** [comment]
Looking at SDL's CI logs, `Windows (clang32|64)` tests have `-- Performing Test HAVE_NO_UNDEFINED - Success`
but `Windows (clang-cl ...)` tests have `-- Performing Test HAVE_NO_UNDEFINED - Failed`

**Comment by AlpRabbit** [comment]
oh, sorry I made a mistake. This one is true.
```yaml
    kind: "try_compile-v1"
    backtrace:
      - "C:/Program Files/CMake/share/cmake-3.26/Modules/Internal/CheckSourceCompiles.cmake:101 (try_compile)"
      - "C:/Program Files/CMake/share/cmake-3.26/Modules/CheckCSourceCompiles.cmake:76 (cmake_check_source_compiles)"
      - "subproject/SDL_image/cmake/PrivateSdlFunctions.cmake:257 (check_c_source_compiles)"
      - "subproject/SDL_image/cmake/PrivateSdlFunctions.cmake:266 (sdl_check_linker_flag)"
      - "subproject/SDL_image/CMakeLists.txt:291 (sdl_target_link_options_no_undefined)"
    checks:
      - "Performing Test HAVE_WL_NO_UNDEFINED"
    directories:
      source: "C:/Users/RYOSU/Projects/sdltest/build/CMakeFiles/CMakeScratch/TryCompile-imv633"
      binary: "C:/Users/RYOSU/Projects/sdltest/build/CMakeFiles/CMakeScratch/TryCompile-imv633"
    cmakeVariables:
      CMAKE_C_FLAGS: ""
      CMAKE_MODULE_PATH: "C:/Users/RYOSU/Projects/sdltest/subproject/SDL_image/cmake"
      CMAKE_POSITION_INDEPENDENT_CODE: "ON"
    buildResult:
      variable: "HAVE_WL_NO_UNDEFINED"
      cached: true
      stdout: |
        Change Dir: C:/Users/RYOSU/Projects/sdltest/build/CMakeFiles/CMakeScratch/TryCompile-imv633
        
        Run Build Command(s):C:/PROGRA~1/ninja/bin/ninja.exe -v cmTC_5848c && [1/2] C:\\PROGRA~1\\LLVM\\bin\\clang.exe -DHAVE_WL_NO_UNDEFINED  -O0 -D_DEBUG -D_DLL -D_MT -Xclang --dependent-lib=msvcrtd -g -Xclang -gcodeview -MD -MT CMakeFiles/cmTC_5848c.dir/src.c.obj -MF CMakeFiles\\cmTC_5848c.dir\\src.c.obj.d -o CMakeFiles/cmTC_5848c.dir/src.c.obj -c C:/Users/RYOSU/Projects/sdltest/build/CMakeFiles/CMakeScratch/TryCompile-imv633/src.c
        [2/2] cmd.exe /C "cd . && C:\\PROGRA~1\\LLVM\\bin\\clang.exe -fuse-ld=lld-link -nostartfiles -nostdlib -O0 -D_DEBUG -D_DLL -D_MT -Xclang --dependent-lib=msvcrtd -g -Xclang -gcodeview -Xlinker /subsystem:console  -Wl,--no-undefined CMakeFiles/cmTC_5848c.dir/src.c.obj -o cmTC_5848c.exe -Xlinker /MANIFEST:EMBED -Xlinker /implib:cmTC_5848c.lib -Xlinker /pdb:cmTC_5848c.pdb -Xlinker /version:0.0   -lkernel32 -luser32 -lgdi32 -lwinspool -lshell32 -lole32 -loleaut32 -luuid -lcomdlg32 -ladvapi32 -loldnames  && cd ."
        lld-link: warning: ignoring unknown argument '--no-undefined'\x0d
        
      exitCode: 0
```

**Comment by AlpRabbit** [comment]
And the result of `cmake --build . --verbose`
```cmd
[504/505] cmd.exe /C "cd . && C:\PROGRA~1\LLVM\bin\clang.exe -fuse-ld=lld-link -nostartfiles -nostdlib -O0 -D_DEBUG -D_DLL -D_MT -Xclang --dependent-lib=msvcrtd -g -Xclang -gcodeview  -Wl,--no-undefined -shared -o subproject\SDL_image\SDL2_imaged.dll  -Xlinker /MANIFEST:EMBED -Xlinker /implib:subproject\SDL_image\SDL2_imaged.lib -Xlinker /pdb:subproject\SDL_image\SDL2_imaged.pdb -Xlinker /version:0.700 subproject/SDL_image/CMakeFiles/SDL2_image.dir/IMG.c.obj subproject/SDL_image/CMakeFiles/SDL2_image.dir/IMG_WIC.c.obj subproject/SDL_image/CMakeFiles/SDL2_image.dir/IMG_avif.c.obj subproject/SDL_image/CMakeFiles/SDL2_image.dir/IMG_bmp.c.obj subproject/SDL_image/CMakeFiles/SDL2_image.dir/IMG_gif.c.obj subproject/SDL_image/CMakeFiles/SDL2_image.dir/IMG_jpg.c.obj subproject/SDL_image/CMakeFiles/SDL2_image.dir/IMG_jxl.c.obj subproject/SDL_image/CMakeFiles/SDL2_image.dir/IMG_lbm.c.obj subproject/SDL_image/CMakeFiles/SDL2_image.dir/IMG_pcx.c.obj subproject/SDL_image/CMakeFiles/SDL2_image.dir/IMG_png.c.obj subproject/SDL_image/CMakeFiles/SDL2_image.dir/IMG_pnm.c.obj subproject/SDL_image/CMakeFiles/SDL2_image.dir/IMG_qoi.c.obj subproject/SDL_image/CMakeFiles/SDL2_image.dir/IMG_stb.c.obj subproject/SDL_image/CMakeFiles/SDL2_image.dir/IMG_svg.c.obj subproject/SDL_image/CMakeFiles/SDL2_image.dir/IMG_tga.c.obj subproject/SDL_image/CMakeFiles/SDL2_image.dir/IMG_tif.c.obj subproject/SDL_image/CMakeFiles/SDL2_image.dir/IMG_webp.c.obj subproject/SDL_image/CMakeFiles/SDL2_image.dir/IMG_xcf.c.obj subproject/SDL_image/CMakeFiles/SDL2_image.dir/IMG_xpm.c.obj subproject/SDL_image/CMakeFiles/SDL2_image.dir/IMG_xv.c.obj subproject/SDL_image/CMakeFiles/SDL2_image.dir/version.rc.res  subproject/sdl/SDL2d.lib  -lkernel32 -luser32 -lgdi32 -lwinspool -lshell32 -lole32 -loleaut32 -luuid -lcomdlg32 -ladvapi32 -loldnames  && cd ."
lld-link: warning: ignoring unknown argument '--no-undefined'
```

**Comment by madebr** [comment]
Thanks! Looks like [SDL does have a clang-cl exception as well](https://github.com/libsdl-org/SDL/blob/8c852f4dcffb118ecdd5d62c7e1b8656733281ce/CMakeLists.txt#L668-L674)

As an experiment, can you replace `-Wl,--no-undefined` with `-Xclang=--Wl,-no-undefined` and see whether this also prints a warning?

**Comment by AlpRabbit** [comment]
I've replaced '-Wl,--no-undefined` with `-Xclang=--Wl,-no-undefined`
```
[1/1] cmd.exe /C "cd . && C:\PROGRA~1\LLVM\bin\clang.exe -fuse-ld=lld-link -nostartfiles -nostdlib -O0 -D_DEBUG -D_DLL -D_MT -Xclang --dependent-lib=msvcrtd -g -Xclang -gcodeview  -Xclang=--Wl,-no-undefined -shared -o subproject\SDL_image\SDL2_imaged.dll  -Xlinker /MANIFEST:EMBED -Xlinker /implib:subproject\SDL_image\SDL2_imaged.lib -Xlinker /pdb:subproject\SDL_image\SDL2_imaged.pdb -Xlinker /version:0.700 subproject/SDL_image/CMakeFiles/SDL2_image.dir/IMG.c.obj subproject/SDL_image/CMakeFiles/SDL2_image.dir/IMG_WIC.c.obj subproject/SDL_image/CMakeFiles/SDL2_image.dir/IMG_avif.c.obj subproject/SDL_image/CMakeFiles/SDL2_image.dir/IMG_bmp.c.obj subproject/SDL_image/CMakeFiles/SDL2_image.dir/IMG_gif.c.obj subproject/SDL_image/CMakeFiles/SDL2_image.dir/IMG_jpg.c.obj subproject/SDL_image/CMakeFiles/SDL2_image.dir/IMG_jxl.c.obj subproject/SDL_image/CMakeFiles/SDL2_image.dir/IMG_lbm.c.obj subproject/SDL_image/CMakeFiles/SDL2_image.dir/IMG_pcx.c.obj subproject/SDL_image/CMakeFiles/SDL2_image.dir/IMG_png.c.obj subproject/SDL_image/CMakeFiles/SDL2_image.dir/IMG_pnm.c.obj subproject/SDL_image/CMakeFiles/SDL2_image.dir/IMG_qoi.c.obj subproject/SDL_image/CMakeFiles/SDL2_image.dir/IMG_stb.c.obj subproject/SDL_image/CMakeFiles/SDL2_image.dir/IMG_svg.c.obj subproject/SDL_image/CMakeFiles/SDL2_image.dir/IMG_tga.c.obj subproject/SDL_image/CMakeFiles/SDL2_image.dir/IMG_tif.c.obj subproject/SDL_image/CMakeFiles/SDL2_image.dir/IMG_webp.c.obj subproject/SDL_image/CMakeFiles/SDL2_image.dir/IMG_xcf.c.obj subproject/SDL_image/CMakeFiles/SDL2_image.dir/IMG_xpm.c.obj subproject/SDL_image/CMakeFiles/SDL2_image.dir/IMG_xv.c.obj subproject/SDL_image/CMakeFiles/SDL2_image.dir/version.rc.res  subproject/sdl/SDL2d.lib  -lkernel32 -luser32 -lgdi32 -lwinspool -lshell32 -lole32 -loleaut32 -luuid -lcomdlg32 -ladvapi32 -loldnames  && cd ."
clang: warning: argument unused during compilation: '-Xclang=--Wl,-no-undefined' [-Wunused-command-line-argument]
```

**Comment by madebr** [comment]
Thanks. I suppose it makes sense this option does not exist on Windows. Missing symbols cannot exist there.

File: cmake/PrivateSdlFunctions.cmake [hunk 1]
```diff
@@ -264,7 +264,7 @@ function(sdl_target_link_options_no_undefined TARGET)
             target_link_options(${TARGET} PRIVATE "-Wl,-undefined,error")
         else()
             sdl_check_linker_flag("-Wl,--no-undefined" HAVE_WL_NO_UNDEFINED)
-            if(HAVE_WL_NO_UNDEFINED)
+            if(HAVE_WL_NO_UNDEFINED AND NOT ((CMAKE_C_COMPILER_ID MATCHES "Clang") AND WINDOWS))
```
Comment (madebr) [inline cmake/PrivateSdlFunctions.cmake:267]: `WINDOWS` is not set by `SDL_image`'s cmake script.
```suggestion
            if(HAVE_WL_NO_UNDEFINED AND NOT ((CMAKE_C_COMPILER_ID MATCHES "Clang") AND WIN32))
```

Comment (AlpRabbit) [reply cmake/PrivateSdlFunctions.cmake]: sorry. I fixed it.

**Review by madebr — APPROVED** [review]

**Pull request merged**